polygon/p2p/message_observer.go

```diff
@@ -1,11 +1,32 @@
 package p2p
 
+import "context"
+
 type MessageObserver[M any] interface {
 	Notify(M)
 }
 
-type ChanMessageObserver[M any] chan M
+func NewChanMessageObserver[M any](ctx context.Context) ChanMessageObserver[M] {
+	return ChanMessageObserver[M]{
+		ctx:     ctx,
+		channel: make(chan M),
+	}
+}
+
+type ChanMessageObserver[M any] struct {
+	ctx     context.Context
+	channel chan M
+}
 
 func (cmo ChanMessageObserver[M]) Notify(msg M) {
```

Comment:
I'd rather pass ctx to Notify as parameter, because at the point of observer creation it might not be available yet.